nova/virt/vmwareapi/ds_util.py

```diff
@@ -443,29 +431,35 @@ def get_sub_folders(session, ds_browser, ds_path):
     return set()
 
 
-def _filter_datastores_matching_storage_policy(session, data_stores,
+def _filter_datastores_matching_storage_policy(session, datastores,
                                                storage_policy):
     """Get datastores matching the given storage policy.
 
-    :param data_stores: the list of retrieve result wrapped datastore objects
+    :param datastores: an iterator over datastore results
     :param storage_policy: the storage policy name
-    :return: the list of datastores conforming to the given storage policy
+    :return: an iterator to datastores conforming to the given storage policy
     """
     profile_id = pbm.get_profile_id_by_name(session, storage_policy)
-    if profile_id:
-        factory = session.pbm.client.factory
-        ds_mors = [oc.obj for oc in data_stores.objects]
-        hubs = pbm.convert_datastores_to_hubs(factory, ds_mors)
-        matching_hubs = pbm.filter_hubs_by_profile(session, hubs,
-                                                   profile_id)
-        if matching_hubs:
-            matching_ds = pbm.filter_datastores_by_hubs(matching_hubs,
-                                                        ds_mors)
-            object_contents = [oc for oc in data_stores.objects
-                               if oc.obj in matching_ds]
-            data_stores.objects = object_contents
-            return data_stores
-    LOG.error("Unable to retrieve storage policy with name %s", storage_policy)
+    if not profile_id:
+        LOG.error("Unable to retrieve storage policy with name %s", storage_policy)
+        return
+
+    factory = session.pbm.client.factory
+
+    # The hub-id is the moref-value of the datastore
+    ds_mors = []
+    ref_to_oc = {}
+    for oc in datastores:
+        ds_mors.append(oc.obj)
+        ref_to_oc[vim_util.get_moref_value(oc.obj)] = oc
+
+    hubs = pbm.convert_datastores_to_hubs(factory, ds_mors)
+    matching_hubs = pbm.filter_hubs_by_profile(session, hubs,
+                                                profile_id)
+    if matching_hubs:
+        # Now we have to map back all the matching ones
+        for hub in matching_hubs:
```

Comment:
Were we talking about the same `if`? I mean this in line 459, but I also like the early return we have now.

Reply:
Ah, I was assuming it was done, because it could return `None`, but it is supposed to return a list.
I'll change that